### PR TITLE
Add support for nRF Connect SDK

### LINKED ISSUE #1493: Add support for nRF Connect SDK
#### Problem
Add support for [nRF Connect SDK](https://www.nordicsemi.com/Software-and-tools/Software/nRF-Connect-SDK).

Currently, it's possible to build and run the door-lock example on Nordic nRF52xx development boards, but the platform layer is implemented using nRF5 SDK. As Nordic Semiconductor we would like to give users the opportunity to use more modern nRF Connect SDK built on top of Zephyr RTOS.

 #### Proposed Solution
* new device layer _nrfconnect_ using nRF Connect SDK / Zephyr API
* necessary changes to system and inet layers to workaround lack of certain features in Zephyr, e.g. pipe(), various socket options or _ifaddrs.h_ interface for IP address enumeration.
* door-lock example ported to the _nrfconnect_ platform

### PR DESCRIPTION
**Pull request by Damian-Nordic**
#### Problem
Add support for nRF Connect SDK.

Currently, it's possible to build and run the door-lock example on Nordic nRF52xx development boards, but the platform layer is implemented using nRF5 SDK. As Nordic Semiconductor we would like to give users the opportunity to use more modern nRF Connect SDK built on top of Zephyr RTOS.

 #### Summary of Changes
* new device layer _nrfconnect_ using nRF Connect SDK / Zephyr API
* necessary changes to system and inet layers to workaround lack of certain features in Zephyr, e.g. pipe(), various socket options or ifaddrs.h interface for IP address enumeration.
* door-lock example ported to the _nrfconnect_ platform

The door-lock example was tested to work correctly over Thread network. Note that functions for BLE and non-volatile storage are still unimplemented and will be added in future PRs. And so will definition of the NCS docker image which may be used in a github workflow.

fixes #1493

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@andy31415 @robszewczyk @jelderton @hawk248 @saurabhst @BroderickCarlin ?

**Review by andy31415 — APPROVED** [review]

**Review by hawk248 — APPROVED** [review]
👍

**Comment by Damian-Nordic** [comment]
FYI, I've just rebased the PR and fixed build errors

File: src/system/SystemConfig.h [hunk 1]
```diff
@@ -609,4 +609,48 @@ struct LwIPEvent;
...
+ *  @brief
+ *      Include missing functions for Zephyr sockets
+ *
+ *  Zephyr socket API lacks some of the functions required by CHIP, e.g. getsockname, recvmsg.
+ *
+ *  By default, provide the necessary functions on Zephyr platforms using sockets
+ */
+#ifndef CHIP_SYSTEM_CONFIG_USE_ZEPHYR_SOCKET_HACKS
+#if CHIP_SYSTEM_CONFIG_USE_SOCKETS && __ZEPHYR__
+#define CHIP_SYSTEM_CONFIG_USE_ZEPHYR_SOCKET_HACKS 1
```
Comment (rwalker-apple) [inline src/system/SystemConfig.h:624]: this characterization probably shouldn't be here

File: src/system/SystemLayer.am [hunk 2]
```diff
@@ -28,6 +28,7 @@ CHIP_BUILD_SYSTEM_LAYER_SOURCE_FILES                  = \
     @top_builddir@/src/system/SystemLayer.cpp           \
     @top_builddir@/src/system/SystemMutex.cpp           \
     @top_builddir@/src/system/SystemObject.cpp          \
+    @top_builddir@/src/system/SystemWakeEvent.cpp       \
```
Comment (rwalker-apple) [inline src/system/SystemLayer.am:31]: please separate adding this functionality into a separate PR that includes tests for the module

File: src/inet/ZephyrSocketHacks.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+{
+    return zsock_sendmsg(sock, msg, flags);
+}
+
+static inline ssize_t recvmsg(int sock, struct msghdr * msg, int flags)
+{
+    // Zephyr doesn't implement recvmsg at all, but if the message vector size is 1 we can simply
+    // translate recvmsg to recvfrom (although we don't get destination interface info in such a case).
+
+    if (msg->msg_iovlen != 1)
```
Comment (rwalker-apple) [inline src/inet/ZephyrSocketHacks.h:44]: this function would also be ok if it only filled one vector, right?  asking all callers to know that this implementation is limited to a vec of len 1 is brittle.

File: src/inet/ZephyrSocketHacks.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ */
+
+/**
+ *    @file
+ *      This header file defines BSD socket API functions which for various
+ *      reasons have not (yet) been implemented or exposed in Zephyr.
+ */
+
+#ifndef INET_ZEPHYR_SOCKET_HACKS_H
+#define INET_ZEPHYR_SOCKET_HACKS_H
```
Comment (rwalker-apple) [inline src/inet/ZephyrSocketHacks.h:25]: Please remove the word "hacks" from this PR.  If there is an issue with this implementation that needs to be addressed at some point, please file an issue.

**Review by rwalker-apple — CHANGES_REQUESTED** [review]

File: src/system/SystemConfig.h (hunk 1 above)
Comment (Damian-Nordic) [reply src/system/SystemConfig.h:624]: Ok, could you suggest an alternative and explain why?

File: src/system/SystemLayer.am (hunk 2 above)
Comment (Damian-Nordic) [reply src/system/SystemLayer.am:31]: ok, can do it

File: src/inet/ZephyrSocketHacks.h (hunk 3 above)
Comment (Damian-Nordic) [reply src/inet/ZephyrSocketHacks.h:44]: yes, but it's similar to other functions in Zephyr which seem to conform to POSIX/BSD API, but only implement part of their functionality, so such things may happen. Unfortunately, there's no proper recvmsg in Zephyr and although we may work on adding it there, for now we may do what I did or add another set of #ifdefs in Inet layer (which I wanted to avoid). If callers handle errors properly an incorrect use of the function should be spotted quickly.

File: src/inet/ZephyrSocketHacks.h (hunk 4 above)
Comment (Damian-Nordic) [reply src/inet/ZephyrSocketHacks.h:25]: would "extensions" be acceptable?

File: src/system/SystemLayer.am (hunk 2 above)
Comment (Damian-Nordic) [reply src/system/SystemLayer.am:31]: see #1594

File: src/inet/ZephyrSocketHacks.h (hunk 3 above)
Comment (rwalker-apple) [reply src/inet/ZephyrSocketHacks.h:44]: what you're saying (I think) is that we should scrub CHIP for calls to recvmsg() with more than one iov?  in such a case, we should just drop any use of recvmsg().

the alternative (which I propose above) is to handle a multi-iov call by only filling the first iov, which is legal for the definition of recvmsg(), right?

File: src/inet/ZephyrSocketHacks.h (hunk 4 above)
Comment (rwalker-apple) [reply src/inet/ZephyrSocketHacks.h:25]: I think if you just dropped the word "HACKS_" from the symbol this would still work as intended, and would not characterize either side (CHIP or Zephyr) as needing a fix.

File: src/inet/ZephyrSocketHacks.h (hunk 3 above)
Comment (Damian-Nordic) [reply src/inet/ZephyrSocketHacks.h:44]: I think multi-iov is not a reason why CHIP uses recvmsg(), I may be wrong though. 

But I like the idea to only fill the first iov. I'll provide a new patch once #1594 is merged.

Anyway, you're right that we may only

**Review by BroderickCarlin — APPROVED** [review]

File: src/inet/ZephyrSocketHacks.h (hunk 3 above)
Comment (Damian-Nordic) [reply src/inet/ZephyrSocketHacks.h]: Fixed in the current version

File: src/inet/ZephyrSocketHacks.h (hunk 4 above)
Comment (Damian-Nordic) [reply src/inet/ZephyrSocketHacks.h]: Fixed in the current version

**Review by rwalker-apple — APPROVED** [review]
